testnet/stacks-node/src/event_dispatcher.rs

```diff
@@ -1615,4 +1698,272 @@ mod test {
+            80,             // HTTP port
+            b"{}",          // Example empty JSON body
+            &Url::parse("http://10.255.255.1/").expect("Failed to parse URL"),
+            timeout_duration,
+        );
+
+        // Measure the elapsed time
+        let elapsed_time = start_time.elapsed();
+
+        // Assert that the connection attempt timed out
+        assert!(
+            result.is_err(),
+            "Expected a timeout error, but got {:?}",
+            result
+        );
+        assert_eq!(
+            result.unwrap_err().kind(),
+            std::io::ErrorKind::TimedOut,
+            "Expected a TimedOut error"
+        );
+
+        // Assert that the elapsed time is within an acceptable range
+        assert!(
+            elapsed_time >= timeout_duration,
+            "Timeout occurred too quickly"
+        );
+        assert!(
+            elapsed_time < timeout_duration + Duration::from_secs(1),
+            "Timeout took too long"
+        );
+    }
+
+    #[test]
+    fn test_send_request_timeout() {
+        // Set up a TcpListener that accepts a connection but delays response
+        let listener = TcpListener::bind("127.0.0.1:0").expect("Failed to bind test listener");
+        let addr = listener.local_addr().unwrap();
+
+        // Spawn a thread that will accept the connection and do nothing, simulating a long delay
+        thread::spawn(move || {
+            let (stream, _addr) = listener.accept().unwrap();
+            // Hold the connection open to simulate a delay
+            thread::sleep(Duration::from_secs(10));
+            drop(stream); // Close the stream
+        });
+
+        // Set a timeout shorter than the sleep duration to force a timeout
+        let connection_timeout = Duration::from_secs(2);
+
+        // Attempt to connect, expecting a timeout error
+        let result = send_request(
+            "127.0.0.1",
+            addr.port(),
+            b"{}",
+            &Url::parse("http://127.0.0.1/").unwrap(),
+            connection_timeout,
+        );
+
+        // Assert that the result is an error, specifically a timeout
+        assert!(
+            result.is_err(),
+            "Expected a timeout error, got: {:?}",
+            result
+        );
+
+        if let Err(err) = result {
+            assert_eq!(
+                err.kind(),
+                std::io::ErrorKind::WouldBlock,
+                "Expected TimedOut error, got: {:?}",
+                err
+            );
+        }
+    }
+
+    fn start_mock_server(response: &str, client_done_signal: Receiver<()>) -> String {
+        // Bind to an available port on localhost
+        let listener = TcpListener::bind("127.0.0.1:0").expect("Failed to bind server");
+        let addr = listener.local_addr().unwrap();
+
+        debug!("Mock server listening on {}", addr);
+
+        // Start the server in a new thread
+        let response = response.to_string();
```

Comment:
As a general rule, if a function is going to make an owned buffer from a slice, it's better just to pass the owned buffer to the function, and let the caller do the conversion if needed

```suggestion
    fn start_mock_server(response: String, client_done_signal: Receiver<()>) -> String {
        // Bind to an available port on localhost
        let listener = TcpListener::bind("127.0.0.1:0").expect("Failed to bind server");
        let addr = listener.local_addr().unwrap();

        debug!("Mock server listening on {}", addr);

        // Start the server in a new thread
```

Reply:
Changed in 98b1b3f9e.